packages/agora-rtc-react/src/hooks/tracks.ts

```diff
@@ -349,3 +355,109 @@ export function useRemoteVideoTracks(
   }, [isConnected, resolvedClient, users]);
   return tracks;
 }
+
+/**
+ * a hook can create a local video track
+ * unpublish track on unmount.
+ */
+export function useLocalVideoTrack(
+  ready = true,
+  client?: IAgoraRTCClient,
+): ICameraVideoTrack | null {
+  const isConnected = useIsConnected(client);
+  const [track, setTrack] = useState<ICameraVideoTrack | null>(null);
+
+  useAsyncEffect(async () => {
+    if (isConnected && ready && !track) {
+      setTrack(await AgoraRTC.createCameraVideoTrack());
+    }
+    if (!isConnected) {
+      setTrack(null);
+    }
+  }, [isConnected, ready]);
+  return track;
+}
+
+/**
+ * a hook can create a local audio track
+ * close track on unmount.
+ */
+export function useLocalAudioTrack(
+  ready = true,
+  audioTrackConfig: MicrophoneAudioTrackInitConfig = { ANS: true, AEC: true },
+  client?: IAgoraRTCClient,
+): IMicrophoneAudioTrack | null {
+  const isConnected = useIsConnected(client);
+
+  const [track, setTrack] = useState<IMicrophoneAudioTrack | null>(null);
+
+  useAsyncEffect(async () => {
+    if (isConnected && ready && !track) {
+      setTrack(await AgoraRTC.createMicrophoneAudioTrack(audioTrackConfig));
+    }
+    if (!isConnected) {
+      setTrack(null);
+    }
+  }, [isConnected, ready]);
+  return track;
+}
+
+/**
+ * publish tacks when readyToPublish is true
+ * unpublish on unmount.
+ */
+export function usePublish(
+  tracks: (ILocalTrack | null)[],
+  readyToPublish = true,
+  client?: IAgoraRTCClient,
+): void {
+  const resolvedClient = useRTCClient(client);
+  const isConnected = useIsConnected(client);
+  //maintain an internal ref to track published
+  const pubTracks = useRef<(ILocalTrack | null)[]>([]);
+  tracks = tracks.filter(track => track !== null && track !== undefined);
```

Comment:
`tracks` will always be different on each rendering, so the following `useAsyncEffect` will always run on each rendering. Is that by design?

Reply:
useLocalAudioTrack & useLocalVideoTrack will ensure that the track is not regenerated every time it is rendered until `isConnected` state has changed
so tracks will be same on each rendering.

Tracks will be changed by following possibilities:

- Component on unMount
- useLocalAudioTrack & useLocalVideoTrack recall

Reply:
Yeah I mean even if the `tracks` param is always the same, the `tracks.filter(...)` creates a new array that is different with the original param, so the effect always runs.